src/workerd/jsg/resource.h

```diff
@@ -1047,20 +1058,11 @@ struct DeserializeInvoker<TypeWrapper,
   }
 };
 
-// SFINAE to detect if a type has a static method called `constructor`.
-template <typename T, typename Constructor = decltype(&T::constructor)>
-constexpr bool hasConstructorMethod(T*) {
-  static_assert(!std::is_member_function_pointer_v<Constructor>,
-      "JSG resource type `constructor` member functions must be static.");
-  // TODO(cleanup): Write our own isMemberFunctionPointer and put it in KJ so we don't have to pull
-  //   in <type_traits>. (See the "Motivation" section of Boost.CallableTraits for why I didn't just
-  //   dive in and do it.)
-  return true;
-}
-// SFINAE to detect if a type has a static method called `constructor`.
-constexpr bool hasConstructorMethod(...) {
-  return false;
-}
+// Concept to detect if a type has a static method called `constructor`.
+// The constructor must be a static method, not a non-static member function.
+template <typename T>
+concept HasConstructorMethod =
+    requires { &T::constructor; } && !std::is_member_function_pointer_v<decltype(&T::constructor)>;
```

Comment:
I think this approach is worse – we'd want to have the static_assert so that a non-static constructor means that we get a compile error instead of HasConstructorMethod silently not being met. (Correct me if I'm wrong here due to how SFINAE works)